views/js/review/plugins/navigation/review-panel/panel.js

```diff
@@ -0,0 +1,672 @@
+     */
+    const cssClasses = {
+        collapsible: 'collapsible',
+        expanded: 'expanded',
+        active: 'active'
+    };
+
+    /**
+     * CSS selectors that match some particular elements
+     * @type {Object}
+     */
+    const cssSelectors = {
+        collapsible: `.${cssClasses.collapsible}`,
+        collapsibleLabel: `.${cssClasses.collapsible} > .review-panel-label`,
+        expanded: `.${cssClasses.expanded}`,
+        active: `.${cssClasses.active}`,
+        control: '[data-control]',
+        content: '.review-panel-content',
+        filtersContainer: '.review-panel-filters',
+        filter: '.review-panel-filter',
+        item: '.review-panel-item',
+        header: '.review-panel-header',
+        footer: '.review-panel-footer',
+        score: '.review-panel-score'
+    };
+
+    /**
+     * Finds an element by its control identifier
+     * @param {jQuery} $container
+     * @param {String} id
+     * @returns {jQuery}
+     */
+    const findControl = ($container, id) => $container.find(`[data-control="${id}"]`);
+
+    /**
+     * Finds every expanded collapsible element within the provided target
+     * @param {jQuery} $target
+     * @returns {jQuery}
+     */
+    const findExpanded = $target => $target.find(cssSelectors.collapsible + cssSelectors.expanded);
+
+    /**
+     * Gets the icon class for a particular item
+     * @param {reviewPanelItem} item
+     * @returns {String}
+     */
+    const getItemIconCls = item => {
+        if (item.maxScore) {
+            if (item.score === item.maxScore) {
+                return 'item-correct';
+            } else {
+                return 'item-incorrect';
+            }
+        }
+        return 'item-info';
+    };
+
+    /**
+     * Reduces an array to another array
+     * @param {Array} array
+     * @param {Function} callback
+     * @returns {Array}
+     */
+    const reduceArray = (array, callback) => (array || []).reduce(callback, []);
+
+    /**
+     * Refines the data to display the panel with respect to the current filter
+     * @param {reviewPanelMap} map
+     * @param {Function} [filter] - A callback function applied to filter the data
+     * @returns {reviewPanelData}
+     */
+    const filterData = (map, filter) => {
+        let score = 0;
+        let maxScore = 0;
+        const itemsMap = new Map();
+        if (!_.isFunction(filter)) {
+            filter = () => true;
+        }
+        return {
+            testMap: {
+                parts: reduceArray(map && map.parts, (testParts, testPart) => {
+                    const sections = reduceArray(testPart.sections, (partSections, partSection) => {
+                        const items = reduceArray(partSection.items, (sectionItems, sectionItem) => {
+                            sectionItem = Object.assign({}, sectionItem);
+                            sectionItem.cls = getItemIconCls(sectionItem);
+
+                            score += sectionItem.score || 0;
```

Comment:
I am not sure, but do we need to calculate score and maxScore every switch between filters? 
Because they don't change. Maybe better calculate them once in setData function?

Reply:
You're right. The score is static and could be computed only once. But to simplify the design and avoid to implement the same loops on several places, I preferred to keep that here.